lib/cuda/cudaCopyNtoRingbuffer.cpp

```diff
@@ -0,0 +1,196 @@
+            }
+        }
+    }
+}
+
+int cudaCopyNtoRingbuffer::wait_on_precondition() {
+    // Wait for all input buffers to have data
+    for (size_t i = 0; i < in_buffers.size(); ++i) {
+        Buffer* buf = in_buffers[i];
+        DEBUG("Waiting for input data frame {} from buffer {}", gpu_frame_id, buf->buffer_name);
+        uint8_t* frame = buf->wait_for_full_frame(unique_name, gpu_frame_id % buf->num_frames);
+        if (!frame)
+            return -1;
+        DEBUG("Input data frame {} is now available from buffer {}", gpu_frame_id,
+              buf->buffer_name);
+    }
+
+    // Wait for ringbuffer space
+    DEBUG("Waiting for ringbuffer space for writing to become available ({} bytes)",
+          total_input_size);
+    std::optional<size_t> val =
+        signal_buffer->wait_for_writable(unique_name, instance_num, total_input_size);
+    DEBUG("Ringbuffer space for writing is now available");
+    if (!val.has_value())
+        return -1;
+    output_cursor = val.value();
+    return 0;
+}
+
+cudaEvent_t cudaCopyNtoRingbuffer::execute(cudaPipelineState& pipestate,
+                                           const std::vector<cudaEvent_t>& pre_events) {
+    (void)pipestate;
+    pre_execute();
+
+    void* rb_memory = device.get_gpu_memory(_gpu_mem_output, signal_buffer->size);
+
+    // Compute total size and wrapping
+    size_t start = output_cursor % signal_buffer->size;
+
+    record_start_event();
+
+    // Merge all input buffers into the ringbuffer sequentially
+    size_t offset = 0;
+    std::vector<std::shared_ptr<chordMetadata>> metas;
+    for (size_t i = 0; i < in_buffers.size(); ++i) {
+        Buffer* buf = in_buffers[i];
+        int buf_index = gpu_frame_id % buf->num_frames;
+        void* host_memory_frame = (void*)buf->frames[buf_index];
+        size_t sz = buf->frame_size;
+
+        // Copy to ringbuffer, handle wrapping if needed
+        if (start + offset + sz <= (size_t)signal_buffer->size) {
+            device.async_copy_host_to_gpu((char*)rb_memory + start + offset, host_memory_frame, sz,
+                                          cuda_stream_id, pre_events[cuda_stream_id], nullptr,
+                                          nullptr);
+        } else {
+            size_t first = signal_buffer->size - (start + offset);
+            size_t second = sz - first;
+            device.async_copy_host_to_gpu((char*)rb_memory + start + offset, host_memory_frame,
+                                          first, cuda_stream_id, pre_events[cuda_stream_id],
+                                          nullptr, nullptr);
+            device.async_copy_host_to_gpu(rb_memory, (char*)host_memory_frame + first, second,
+                                          cuda_stream_id, nullptr, nullptr, nullptr);
+        }
+
+        offset += sz;
+    }
+
+    // We only need to set the metadata on the ring buffer once.
+    if (first_time && instance_num == 0) {
```

Comment:
Thinking about this more we should implement a cross-check of the metadata on every call to this function.